pkg/npm/npm.go

```diff
@@ -9,58 +9,129 @@ import (
-	// `description`, `repository`, and `license` fields in the package.json file.
-	c := exec.Command("npm", "install", "--loglevel=error")
+// app located there. If the `PULUMI_PREFER_YARN` environment variable is set, `yarn install` is used
+// instead of `npm install`.
+func Install(dir string, stdout, stderr io.Writer) (string, error) {
+	c, npm, bin, err := getCmd("install")
+	if err != nil {
+		return bin, err
+	}
 	c.Dir = dir
 
 	// Run the command.
-	c.Stdout = stdout
-	c.Stderr = stderr
-	if err := c.Run(); err != nil {
-		return err
+	if err = runCmd(c, npm, stdout, stderr); err != nil {
+		return bin, err
 	}
 
 	// Ensure the "node_modules" directory exists.
 	if _, err := os.Stat("node_modules"); os.IsNotExist(err) {
-		return errors.Errorf("npm install reported success, but node_modules directory is missing")
+		return bin, errors.Errorf("%s install reported success, but node_modules directory is missing", bin)
+	}
+
+	return bin, nil
+}
+
+// getCmd returns the exec.Cmd used to install NPM dependencies. It will either use `npm` or `yarn` depending
+// on what is available on the current path, and if `PULUMI_PREFER_YARN` is truthy.
+// The boolean return parameter indicates if `npm` is chosen or not (instead of `yarn`).
+func getCmd(command string) (*exec.Cmd, bool, string, error) {
+	if preferYarn() {
+		const file = "yarn"
+		yarnPath, err := exec.LookPath(file)
+		if err == nil {
+			return exec.Command(yarnPath, command), false, file, nil
+		}
+		logging.Warningf("could not find yarn on the $PATH, trying npm instead: %v", err)
+	}
+
+	const file = "npm"
+	npmPath, err := exec.LookPath(file)
+	if err != nil {
+		return nil, false, file, errors.Wrapf(err, "could not find npm on the $PATH; npm is installed with Node.js "+
+			"available at https://nodejs.org/")
+	}
+	// We pass `--loglevel=error` to prevent `npm` from printing warnings about missing
+	// `description`, `repository`, and `license` fields in the package.json file.
+	return exec.Command(npmPath, command, "--loglevel=error"), true, file, nil
+}
+
+// runCmd handles hooking up `stdout` and `stderr` and then runs the command.
+func runCmd(c *exec.Cmd, npm bool, stdout, stderr io.Writer) error {
+	// Setup `stdout` and `stderr`.
+	// `stderr` is ignored when `yarn` is used because it outputs warnings like "package.json: No license field"
+	// to `stderr` that we don't need to show.
+	c.Stdout = stdout
+	var stderrBuffer bytes.Buffer
+	if npm {
+		c.Stderr = stderr
+	} else {
+		c.Stderr = &stderrBuffer
+	}
+
+	// Run the command.
+	if err := c.Run(); err != nil {
+		// If we failed, and we're using `yarn`, write out any bytes that were written to `stderr`.
+		if !npm {
+			stderr.Write(stderrBuffer.Bytes())
+		}
+		return err
 	}
 
 	return nil
 }
+
+// preferYarn returns true if the `PULUMI_PREFER_YARN` environment variable is set.
+func preferYarn() bool {
+	return cmdutil.IsTruthy(os.Getenv("PULUMI_PREFER_YARN"))
```

Comment:
BTW, I'm open to better naming ideas here. `PULUMI_USE_YARN`?

Reply:
I think "prefer" is fine given the current behavior. If we change this to "use" I would expect us to fail if `yarn` cannot be found.

Reply:
That would be what I would expect as a user, that if I specify to use yarn, it should fail if yarn cannot be found. I would not want it to fallback to npm.